### PR TITLE
:construction: OpenMP compilation with set

### PR DESCRIPTION
**Pull request by kks32**
**Describe the PR**
Add appropriate flags in OpenMP to enable parallelization on C++ sets

**Related Issues/PRs**
https://cb-geo.discourse.group/t/openmp-compilation-issue/185/2

**Additional context**

OpenMP fails with: 
```
error: invalid controlling predicate
     for (auto sitr = set.begin(); sitr != set.cend(); ++sitr) {
     ^~~
```

However, this error does not occur on newer versions of OpenMP/GCC or in Fedora. The error seems to be specific to older versions of Ubuntu.

### CODE REVIEW

**Comment by kks32** [comment]
@jgiven100 and @ezrayst could you please compile on your Ubuntu machines and let me know if this problem is resolved?

**Comment by jgiven100** [comment]
I am unsure about @ezrayst but my initial `error: invalid controlling predicate` issue occured for most (if not all) instances of the following language in `mesh.tcc`:
```
#pragma omp parallel for
  for (auto...
```
Although I no longer get `error: invalid controlling predicate` associated with [this commit](https://github.com/cb-geo/mpm/pull/668/commits/d7c284c98f57f7a8ec7a85100b9dcc295ef1a2aa) in `mesh.tcc`, it is still occirng for lines:
* 73
* 81
* 176
* etc...

**Comment by jgiven100** [comment]
I am unsure about @ezrayst  but my initial `error: invalid controlling predicate` issue occured for most (if not all) instances of the following language in `mesh.tcc`:
```
#pragma omp parallel for
  for (auto...
```
Although I no longer get `error: invalid controlling predicate` associated with the lines in [this commit](https://github.com/cb-geo/mpm/pull/668/commits/d7c284c98f57f7a8ec7a85100b9dcc295ef1a2aa) in `mesh.tcc`, it is still occirng for lines:
* 73
* 81
* 176
* etc...

**Comment by kks32** [comment]
@jgiven100 Thanks, looks like OpenMP 2015 standard in Ubuntu 18.04. Please try the following to update gcc and recompile:

```
sudo add-apt-repository -y ppa:ubuntu-toolchain-r/test
sudo apt install gcc-9 g++-9
sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-9 90 --slave /usr/bin/g++ g++ /usr/bin/g++-9 --slave /usr/bin/gcov gcov /usr/bin/gcov-9
```

**Comment by jgiven100** [comment]
:+1: I can successfully build & run `mpmtest`
Thanks, @kks32!

**Review by jgiven100 — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -56,14 +57,19 @@ dnf install -y boost boost-devel clang clang-analyzer clang-tools-extra cmake cp
...
 
 ```
-sudo apt-get install -y gcc git libboost-all-dev libeigen3-dev libhdf5-serial-dev libopenmpi-dev
+sudo apt update
+sudo apt upgrade
+sudo apt install -y gcc git libboost-all-dev libeigen3-dev libhdf5-serial-dev libopenmpi-dev libomp-dev
+sudo apt install software-properties-common
+sudo add-apt-repository -y ppa:ubuntu-toolchain-r/test
+sudo apt install gcc-9 g++-9
+sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-9 90 --slave /usr/bin/g++ g++ /usr/bin/g++-9 --slave /usr/bin/gcov gcov /usr/bin/gcov-9
```
Comment (bodhinandach) [inline README.md]: Can we say something on the version of Ubuntu? I mean I can live with the original installation instruction with 20.04, i.e.

`sudo apt-get install -y gcc git libboost-all-dev libeigen3-dev libhdf5-serial-dev libopenmpi-dev`

 but I think many people is still using 18.04.

**Comment by ezrayst** [comment]
I could compile and passed all tests too! Thanks @kks32

File: README.md (hunk 1 above)
Comment (ezrayst) [reply README.md]: Maybe it's better to say minimum gcc v9?

**Review by ezrayst — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (kks32) [reply README.md]: We don't need GCC version 9, we merely need OpenMP to be more recent.

**Review by bodhinandach — APPROVED** [review]
Thanks! This looks amazing.

**Pull request merged**